src/librustc_mir/transform/simplify.rs

```diff
@@ -367,9 +367,14 @@ impl<'a, 'tcx> Visitor<'tcx> for DeclMarker<'a, 'tcx> {
                 if let StatementKind::Assign(
                     box (p, Rvalue::Use(Operand::Constant(c)))
                 ) = &stmt.kind {
-                    if !p.is_indirect() {
-                        trace!("skipping store of const value {:?} to {:?}", c, p);
-                        return;
+                    match c.literal.val {
```

Comment:
this is the regression fix for the regression introduced in https://github.com/rust-lang/rust/pull/66216

Reply:
Looks like a rather unprincipled hack to me -- why only constants and why specifically *unevaluated* constants? Cc @eddyb

Reply:
because all other constants can't hide errors in their body. This bug is solely about constants *never* getting evaluated, thus we are not getting any error

Reply:
That seems quite annoying that we cannot DCE uses of constants -- and quite fragile, too.

Reply:
It's not quite dead code. Just like `let _: () = panic!();` still panics, we should still error on `let _: () = ERRONEOUS_CONSTANT;`

Reply:
Just because a constant's value is `()`, it does not mean that evaluating the constant can't have the "compilation fails" side effect.

Reply:
I think that is disputable... this makes consts more side-effecting than one might think.

Not being able to replace, in MIR optimizations, the computation of any ZST by just a ZST scalar is a big foot-gun, and I expect this to be quite detrimental to MIR-level optimizations. I feel like we should look into ways to obtain the "compilation failure" effect without constraining MIR transformations. For example, a MIR body could come with a list of consts that need to be error-free for this function to be compilable; that would avoid the need to keep these consts around in the MIR itself.

Reply:
Oh yea that would be a much cleaner way. Do you want me to rewrite the PR with that or can we merge it in the hope of getting it into beta before it becomes stable?

Reply:
I don't want to block anything here, certainly not a soundness fix for a release happening next week. ;)
So, sure, go ahead. Maybe add a link to this discussion in the code for people that stumble upon this and wonder what is going on and why.

Reply:
> For example, a MIR body could come with a list of consts that need to be error-free for this function to be compilable; that would avoid the need to keep these consts around in the MIR itself.

This is tempting me to use indices into such a list to refer to consts from the MIR itself.
Even if `ty::Const` is just a pointer, this would allow lowering that to `u32`, so maybe there's some place where we could take advantage of that?

I guess this is more of an unrelated "something @oli-obk / @spastorino could be doing later" thing?

Reply:
:joy: 

> This is tempting me to use indices into such a list to refer to consts from the MIR itself.
Even if ty::Const is just a pointer, this would allow lowering that to u32, so maybe there's some place where we could take advantage of that?

omg yes